### PR TITLE
Determine set of publicly accessible snapshots via TDR (#2978)

### PR DESCRIPTION
**Pull request by noah-aviel-dove**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [ ] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [ ] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in sandbox and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [ ] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: environment.py
Comment (hannes-ucsc) [inline environment.py:289]: Dash missing. 

```suggestion
        'AZUL_GOOGLE_SERVICE_ACCOUNT_PUBLIC': 'azul-ucsc-{AZUL_DEPLOYMENT_INCARNATION}-public-{AZUL_DEPLOYMENT_STAGE}',
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I wanted to do a WIP review but the staleness and conflicts got in the way. You should try to keep PRs up-to-date, even while they're in review.

**Comment by noah-aviel-dove** [comment]
Retracting review request because the upstream PR has changed.

**Comment by hannes-ucsc** [comment]
I'm not gonna look at these until the base PR has been approved.

**Review by danielsotirhos — APPROVED** [review]
LGTM

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: scripts/register_sam.py
Comment (hannes-ucsc) [inline scripts/register_sam.py:47]: I don't think this necessarily constitutes and error. One of the SA should be able to access snapshots. Both don't have to.

Comment (hannes-ucsc) [reply scripts/register_sam.py:47]: "public_tdr service account" makes no sense given the nomenclature used in documentation and configuration. Use the established conceptual name or the actual name to avoid the confusion altogether. Please also look for places where we need to log the SA name since it is now ambiguous as to which account we're talking about. SAMClient.register_with_sam comes to mind for example.

File: scripts/register_sam.py
Comment (hannes-ucsc) [inline scripts/register_sam.py:49]: Be specific, "over-authorized" is a bit vague. And log the difference.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Comment by amarjandu** [comment]
changed base for this PR due to #3023 being merged into develop.

File: environment.py
Comment (noah-aviel-dove) [reply environment.py:281]: I misinterpreted the suggested diff(s) at the top of this conversation.

File: src/azul/plugins/__init__.py
Comment (noah-aviel-dove) [reply src/azul/plugins/__init__.py]: > This is a an odd signature.

I agree, but from the controller's perspective, all four combinations of `None`/`not None` parameters and return values do occur: 

- `None` in, `None` out: unauthenticated request to a DSS catalog
- `None` in, `str` out: unauthenticated request to a TDR catalog (see next paragraph)
- `Authentication` in, `None` out: authenticated request to a DSS catalog
- `Authentication` in, `str` out: authenticated request to a TDR catalog

> If None is passed, what, other than None, can this function return?

If `None` is passed to the TDR plugin's implementation, it returns the email address of the public service account. That's the motivation for this method's existence: the ability to generate a meaningful cache key for the publicly accessible sources in a way that doesn't bypass the repository plugin.

> Would the DSS implementation of it ever be called with anything but None?

Yes. If the current request is authenticated, an instance of `azul.auth.OAuth2` will be passed to this method, regardless of whether it's the DSS or TDR implementation. The `RepositoryController` does not know which plugins accept authentication until it inspects the value returned by this method.

(Just organizing my thoughts, this is clearly a topic for parking lot)

**Comment by noah-aviel-dove** [comment]
Retracting review request to fix a bug I found

File: src/azul/service/repository_controller.py
Comment (hannes-ucsc) [inline src/azul/service/repository_controller.py:261]: Nah, this is back to being being overly complicated. If the issue is not wanting to create many cache lines for the public SA, then we should find a more comprehensible way of addressing that.

Maybe the fall-back to the public service account should be done in the list_sources implementation of the TDR plugin. The cache key should be None when authentication is None and authentication.identity() when it isn't. The default_authentication method would go away in that approach.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/plugins/repository/tdr/__init__.py
Comment (noah-aviel-dove) [reply src/azul/plugins/repository/tdr/__init__.py:182]: It seems more efficient to move this method so that it benefits from the existing explanation for `Plugin._tdr`

File: src/azul/service/repository_controller.py
Comment (hannes-ucsc) [reply src/azul/service/repository_controller.py:261]: Looks like you haven't addressed this yet, which is fine since you haven't asked for review. We do have to accelerate this a bit so let me be more specific and lay this out in a patch:

```diff
Index: src/azul/plugins/repository/tdr/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/repository/tdr/__init__.py b/src/azul/plugins/repository/tdr/__init__.py
--- a/src/azul/plugins/repository/tdr/__init__.py	(revision c41a3ed99ba18f0543b31c964228679b15654069)
+++ b/src/azul/plugins/repository/tdr/__init__.py	(date 1626193217546)
@@ -179,29 +179,27 @@
     def sources(self) -> AbstractSet[TDRSourceSpec]:
         return self._sources
 
-    def default_authentication(self) -> Tuple[Authentication, str]:
-        public_tdr = self._public_authenticated_tdr()
-        return OAuth2(public_tdr.credentials.token), public_tdr.authentication_id()
-
     def list_sources(self,
                      authentication: Optional[Authentication]
                      ) -> List[TDRSourceRef]:
-        if isinstance(authentication, OAuth2):
+        if authentication is None:
+            tdr = TDRClient.with_public_service_account_credentials()
+        elif isinstance(authentication, OAuth2):
             tdr = TDRClient.with_user_credentials(authentication)
-            configured_specs_by_name = {spec.name: spec for spec in self.sources}
-            snapshot_ids_by_name = {
-                name: id
-                for id, name in tdr.snapshot_names_by_id().items()
-                if name in configured_specs_by_name
-            }
-            return [
-                TDRSourceRef(id=id,
-                             spec=configured_specs_by_name[name])
-                for name, id in snapshot_ids_by_name.items()
-            ]
-        else:
-            raise PermissionError('Unsupported authentication format',
-                                  type(authentication))
+        else:
+            raise PermissionError('Unsupported authentication format',
+                                  type(authentication))
+        configured_specs_by_name = {spec.name: spec for spec in self.sources}
+        snapshot_ids_by_name = {
+            name: id
+            for id, name in tdr.snapshot_names_by_id().items()
+            if name in configured_specs_by_name
+        }
+        return [
+            TDRSourceRef(id=id,
+                         spec=configured_specs_by_name[name])
+            for name, id in snapshot_ids_by_name.items()
+        ]
 
     @property
     def tdr(self):
Index: src/azul/plugins/__init__.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/plugins/__init__.py b/src/azul/plugins/__init__.py
--- a/src/azul/plugins/__init__.py	(revision c41a3ed99ba18f0543b31c964228679b15654069)
+++ b/src/azul/plugins/__init__.py	(date 1626194964664)
@@ -196,14 +196,6 @@
         """
         raise NotImplementedError
 
-    def default_authentication(self) -> Union[Tuple[Authentication, str],
-                                              Tuple[None, None]]:
-        """
-        Authentication and authentication identity to use for accessing the
-        remote repository when none is provided by the request.
-        """
-        return None, None
-
     @abstractmethod
     def list_sources(self,
                      authentication: Optional[Authentication]
Index: src/azul/service/repository_controller.py
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/azul/service/repository_controller.py b/src/azul/service/repository_controller.py
--- a/src/azul/service/repository_controller.py	(revision c41a3ed99ba18f0543b31c964228679b15654069)
+++ b/src/azul/service/repository_controller.py	(date 1626193266346)
@@ -251,32 +251,28 @@
         authentication = request.authentication
         plugin = self.repository_plugin(catalog)
 
-        def list_sources() -> Iterable[SourceRef]:
+        cache_key = (
+            catalog,
+            '' if authentication is None else authentication.identity()
+        )
+        joiner = ':'
+        assert not any(joiner in c for c in cache_key), cache_key
+        cache_key = joiner.join(cache_key)
+        try:
+            sources = self._source_cache_service.get(cache_key)
+        except CacheMiss:
             try:
-                return plugin.list_sources(authentication)
+                sources = plugin.list_sources(authentication)
             except PermissionError:
                 raise UnauthorizedError
-
-        if authentication is None:
-            authentication, authentication_identity = plugin.default_authentication()
-        else:
-            authentication_identity = authentication.identity()
-
-        if authentication is None:
-            sources = list_sources()
+            else:
+                sources = [source.to_json() for source in sources]
+                self._source_cache_service.put(cache_key, sources)
         else:
-            cache_key = f'{catalog}:{authentication_identity}'
-            try:
-                source_jsons = self._source_cache_service.get(cache_key)
-            except CacheMiss:
-                sources = list_sources()
-                source_jsons = [source.to_json() for source in sources]
-                self._source_cache_service.put(cache_key, source_jsons)
-            else:
-                sources = [
-                    plugin.source_from_json(source_json)
-                    for source_json in source_jsons
-                ]
+            sources = [
+                plugin.source_from_json(source)
+                for source in sources
+            ]
         return [
             {
                 'sourceId': source.id,
```

Aside from simplifying things, this also removes `source_jsons` and reuses `sources` instead, and has more robust composition of the cache key.

It caches DSS sources but I see no downsides to that. It also wouldn't invalidate the "public" cache line if the public service account changes but that is rare. The current implementation would still let the "public" cache go stale if the permissions of the SA change. The latter is much more likely to occur and we address that by having a short-ish expiration. The short expiration also takes care of a change in the public SA.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
No bump. Unsolicited review.

**Review by hannes-ucsc — APPROVED** [review]

**Comment by hannes-ucsc** [comment]
I will hold this back a little and try to get PRs #3218 and #3225 merged first so we can do the `prod` promotion today.

I'll be ready to review the next PR in this chain (#2954) at your earliest convenience, @noah-aviel-dove, even in chained form (not rebased on develop).

**Comment by hannes-ucsc** [comment]
@amarjandu, I plan to fill in the demo details later.

**Pull request merged**
